File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int l1 = nums1.length;
        int l2 = nums2.length;
        ArrayList<Integer> srtd = new ArrayList<>();
        int a = 0;
        int b = 0;
        while (a < l1 && b < l2) {
            if (nums1[a] < nums2[b]) {
                srtd.add(nums1[a]);
                a++;
            } else if (nums1[a] > nums2[b]) {
                srtd.add(nums2[b]);
                b++;
            } else {
                srtd.add(nums1[a]);
                srtd.add(nums2[b]);
                a++;
                b++;
            }
        }
        while (a < l1) {
            srtd.add(nums1[a]);
            a++;
        }
        while (b < l2) {
            srtd.add(nums2[b]);
            b++;
        }
        double median = 0;
        int end = srtd.size() - 1;
        if (end % 2 != 0) {
            median = (srtd.get(end / 2) + srtd.get((end / 2) + 1)) / (double) 2;
        } else {
            median = (double) srtd.get(end / 2);
        }
        return median;
    }
}

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        return medianOfTwoSorted(nums1, nums2)
    
def medianOfTwoSorted(A, B):
    
    n1, n2 = len(A), len(B)
    if (n1 > n2):
        return medianOfTwoSorted(B, A)
    
    low = 0
    hi = n1
    left = (n1+n2+1) // 2

    while (low<=hi):
         
        midA = (low + hi) // 2
        midB = left - midA

        l1 = A[midA-1] if midA > 0 else float('-inf')
        l2 = B[midB-1] if midB > 0 else float('-inf')
        r1 = A[midA] if midA < len(A) else float('inf')
        r2 = B[midB] if midB < len(B) else float('inf')

        if (l1<=r2 and l2<=r1):
            if (n1+n2) % 2 == 0:
                return (max(l1, l2) + min(r1, r2)) / 2.0
            else:
                return max(l1, l2)
        elif (l1>r2):
            hi = midA - 1
        elif(l2 > r1):
            low = midA + 1